source/extensions/filters/http/rate_limit_quota/client.h

```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/common/pure.h"
+#include "envoy/service/rate_limit_quota/v3/rlqs.pb.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace RateLimitQuota {
+
+using RateLimitQuotaResponsePtr =
+    std::unique_ptr<envoy::service::rate_limit_quota::v3::RateLimitQuotaResponse>;
+
+/**
+ * Async callbacks used during rateLimit() calls.
+ */
+class RateLimitQuotaCallbacks {
+public:
+  virtual ~RateLimitQuotaCallbacks() = default;
+
+  virtual void
+  onReceive(envoy::service::rate_limit_quota::v3::RateLimitQuotaResponse* response) PURE;
```

Comment:
Please use reference if you do not intend to take ownership of the `response`.

Also I would think that a callback to indicate that the stream was closed is also needed, so you can detect error conditions.

Maybe also make method name more descriptive, like `onQuotaResponse`

Reply:
The caller site of this function which inherits from grcp async_client's [onReceiveMessage](https://github.com/envoyproxy/envoy/blob/35bf11a0f14a996d4e95b5302b0a953e0beba9f6/source/common/grpc/typed_async_client.h#L87) that has the smart pointer as function arg. So I just use the raw pointer here by simply via `get()`.

From the code [here](https://github.com/envoyproxy/envoy/blob/35bf11a0f14a996d4e95b5302b0a953e0beba9f6/source/common/grpc/typed_async_client.h#L97), this pointer will not be null. So you are right the reference is slightly better here as it can manifest its `non-nullness`.

Will update it